runtime/interpreter/interpreter_expression.go

```diff
@@ -1142,7 +1184,66 @@ func (interpreter *Interpreter) VisitPathExpression(expression *ast.PathExpressi
 	)
 }
 
-func (interpreter *Interpreter) VisitAttachExpression(_ *ast.AttachExpression) Value {
-	// TODO: implement this
-	return nil
+func (interpreter *Interpreter) VisitAttachExpression(attachExpression *ast.AttachExpression) Value {
+
+	base, ok := interpreter.evalExpression(attachExpression.Base).(*CompositeValue)
+	// we enforce this in the checker
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	locationRange := LocationRange{
+		Location:    interpreter.Location,
+		HasPosition: attachExpression,
+	}
+
+	if inIteration := interpreter.SharedState.inAttachmentIteration[base]; inIteration {
+		panic(AttachmentIterationMutationError{
+			Value:         base,
+			LocationRange: locationRange,
+		})
+	}
```

Comment:
oh, this is no longer valid. I was under the assumption that iteration was/will be removed altogether until the reflection APIs are finalized, which is not the case.

Reply:
Ah, I see, this is a left-over from the iteration API. Agreed, maybe remove this and other iteration-related code that is still in the current PRs.

Reply:
This check is actually still necessary. Although the ability for users to explicitly iterate over attachments with `forEachAttachment` has been removed from this PR, implicit iteration can still happen when a resource containing attachments is `destroy`ed: we iterate over all the attachments on that resource and destroy them one by one. We still need to make sure attachments are not added or removed from that resource during that iteration, as it could cause the order to change and thus one of the attachments (and potential nested resource) to not be destroyed.

Reply:
Added a test demonstrating this case in https://github.com/onflow/cadence/pull/2113/commits/8499a7c79e1bfd5af982a6fe55952d3b74d6861b

Reply:
Ah, right! 👍